satellite/overlay/nodeselectioncache.go

```diff
@@ -0,0 +1,196 @@
+// Copyright (C) 2019 Storj Labs, Incache.
+// See LICENSE for copying information.
+
+package overlay
+
+import (
+	"context"
+	"math/rand"
+	"sync"
+	"time"
+
+	"go.uber.org/zap"
+)
+
+// CacheDB implements the database for overlay node selection cache
+//
+// architecture: Database
+type CacheDB interface {
+	// SelectAllStorageNodesUpload returns all nodes that qualify to store data, organized as reputable nodes and new nodes
+	SelectAllStorageNodesUpload(ctx context.Context, selectionCfg NodeSelectionConfig) (reputable, new []*SelectedNode, err error)
+}
+
+// CacheConfig is a configuration for overlay node selection cache.
+type CacheConfig struct {
+	Staleness time.Duration `help:"how stale the node selection cache can be" releaseDefault:"3m" devDefault:"5m"`
+}
+
+// NodeSelectionCache keeps a list of all the storage nodes that are qualified to store data
+// We organize the nodes by if they are reputable or a new node on the network.
+// The cache will get refreshed once the staleness time has past.
+type NodeSelectionCache struct {
+	log             *zap.Logger
+	db              CacheDB
+	selectionConfig NodeSelectionConfig
+	staleness       time.Duration
+
+	mu   sync.RWMutex
+	data *state
+}
+
+type state struct {
+	lastRefresh time.Time
+
+	mu             sync.RWMutex
+	reputableNodes []*SelectedNode
+	newNodes       []*SelectedNode
+}
+
+// NewNodeSelectionCache creates a new cache that keeps a list of all the storage nodes that are qualified to store data
+func NewNodeSelectionCache(log *zap.Logger, db CacheDB, staleness time.Duration, config NodeSelectionConfig) *NodeSelectionCache {
+	return &NodeSelectionCache{
+		log:             log,
+		db:              db,
+		staleness:       staleness,
+		selectionConfig: config,
+		data:            &state{},
+	}
+}
+
+// Init populates the cache with all of the reputableNodes and newNode nodes
+// that qualify to upload data from the nodes table in the overlay database
+func (cache *NodeSelectionCache) Init(ctx context.Context) (err error) {
+	defer mon.Task()(&ctx)(&err)
+	cache.mu.Lock()
+	defer cache.mu.Unlock()
+	_, err = cache.refresh(ctx)
+	return err
+}
+
+// Refresh calls out to the database and refreshes the cache with the current data
+// from the nodes table, then sets time that the last refresh occurred so we know when
+// to refresh again in the future
+func (cache *NodeSelectionCache) refresh(ctx context.Context) (cachData *state, err error) {
+	defer mon.Task()(&ctx)(&err)
```

Comment:
This code is missing locking cache.mu.Lock. When two concurrent callers call "GetNodes" then this would race.